src/testing/flow_solver_cases/naca0012.cpp

```diff
@@ -0,0 +1,116 @@
+#include "dg/dg.h"
+#include <deal.II/base/table_handler.h>
+#include "mesh/grids/naca_airfoil_grid.hpp"
+#include "mesh/gmsh_reader.hpp"
+#include "functional/lift_drag.hpp"
+
+namespace PHiLiP {
+namespace Tests {
+//=========================================================
+// NACA0012
+//=========================================================
+template <int dim, int nstate>
+NACA0012<dim, nstate>::NACA0012(const PHiLiP::Parameters::AllParameters *const parameters_input)
+        : FlowSolverCaseBase<dim, nstate>(parameters_input)
+{
+}
+
+template <int dim, int nstate>
+void NACA0012<dim,nstate>::display_flow_solver_setup() const
+{
+    using PDE_enum = Parameters::AllParameters::PartialDifferentialEquation;
+    const PDE_enum pde_type = this->all_param.pde_type;
+    const double pi = atan(1.0) * 4.0;
+    std::string pde_string;
+    if (pde_type == PDE_enum::euler)                {pde_string = "euler";}
+    if (pde_type == PDE_enum::navier_stokes)        {pde_string = "navier_stokes";}
+    this->pcout << "- PDE Type: " << pde_string << std::endl;
+    this->pcout << "- Polynomial degree: " << this->all_param.grid_refinement_study_param.poly_degree << std::endl;
+    this->pcout << "- Courant-Friedrich-Lewy number: " << this->all_param.flow_solver_param.courant_friedrich_lewy_number << std::endl;
+    this->pcout << "- Freestream Reynolds number: " << this->all_param.navier_stokes_param.reynolds_number_inf << std::endl;
+    this->pcout << "- Freestream Mach number: " << this->all_param.euler_param.mach_inf << std::endl;
+    this->pcout << "- Angle of attack: " << this->all_param.euler_param.angle_of_attack*180/pi << std::endl;
+    this->pcout << "- Side-slip angle: " << this->all_param.euler_param.side_slip_angle*180/pi << std::endl;
+}
+
+template <int dim, int nstate>
+std::shared_ptr<Triangulation> NACA0012<dim,nstate>::generate_grid() const
+{
+    std::shared_ptr<Triangulation> grid = std::make_shared<Triangulation>(
+            MPI_COMM_WORLD,
+            typename dealii::Triangulation<dim>::MeshSmoothing(
+                    dealii::Triangulation<dim>::smoothing_on_refinement |
+                    dealii::Triangulation<dim>::smoothing_on_coarsening));
+
+    dealii::GridGenerator::Airfoil::AdditionalData airfoil_data;
+    airfoil_data.airfoil_type = "NACA";
+    airfoil_data.naca_id      = "0012";
+    airfoil_data.airfoil_length = 1.0;
+    airfoil_data.height         = 150.0; // Farfield radius.
+    airfoil_data.length_b2      = 150.0;
+    airfoil_data.incline_factor = 0.0;
+    airfoil_data.bias_factor    = 5.0;
+    airfoil_data.refinements    = 0;
+    airfoil_data.n_subdivision_x_0 = 4;
+    airfoil_data.n_subdivision_x_1 = 4;
+    airfoil_data.n_subdivision_x_2 = 4;
+    airfoil_data.n_subdivision_y = 4;
+
+    airfoil_data.airfoil_sampling_factor = 100000; // default 2
+
+    Grids::naca_airfoil(reinterpret_cast<dealii::parallel::distributed::Triangulation<2> &>(*grid), airfoil_data);
```

Comment:
Not entirely sure, there might be something wrong with the templates. The error I'm getting if I don't use the reinterpret_cast is: 

error: invalid initialization of reference of type ‘dealii::parallel::distributed::Triangulation<2>&’ from expression of type ‘std::__shared_ptr_access<dealii::parallel::distributed::Triangulation<3>, __gnu_cxx::_S_atomic, false, false>::element_type’ {aka ‘dealii::parallel::distributed::Triangulation<3>’}

Maybe you have a better idea of why that's happening and how to prevent it?

Reply:
Seems like it thinks `grid` is a 3D Triangulation, but the function only works for 2D Triangulations. The reinterpret cast makes the 3D Triangulation being read as a 2D Triangulation, which is no good either.

I'm suprised it generates this error since you seem to only instantiate this class for PHILIP_DIM==2. Is the comment about moving the "using" helping fix this?

Reply:
No, moving the usings did not help.

Reply:
Can you try `make 2D` and or `make 3D`, and see if the error pops up in both of them? I would expect it to only fail compilation for `make 3D`.

Reply:
It fails only for make 3D. It also fails for make 1D, with a similar error message.

Reply:
How about a:
if constexpr(dim==2) {
    Grids::naca_airfoil()
} else {
    // Error message
}

Reply:
This is an equivalent of compiler #if PHILIP_DIM==2, but prettier. I think that the ideal solution is to allow Grids::naca_airfoil() to take in a generic grid dimension (and fail appropriately). But this solution is fine since we'll never run that flow case for 3D.

Reply:
Using the constexpr does not work and gives the same error message, since it's at compile time. Using the #if preprocessor directive does work. So I guess the solution is:

#if PHILIP_DIM==2
Grids::naca_airfoil()
#endif

If you're good with that, I'll commit it.

Reply:
Sure. Better than than the reinterpret cast. The reinterpret_cast is pretty "evil" and usually done for low-level programming C-style.

Kinda annoyed that it gets a compilation error for 1 and 3 dimension when it shouldn't even be instantiated/compiled.

Reply:
Following Discord conversation, I have replaced the entire function with something simpler, since it's just a dummy triangulation. This also solves the issue here.